### DISCUSSION TITLE
Using yarn in postUpgradeTask

### DISCUSSION THREAD

**Discussion by steknowis** [renovatebot/renovate#14956] [Request Help] [answered]
Hello all,

We have introduced renovate in our company and are very satisfied with the flexibility and possibilities it provides but we have one particular issue which you can might help us with.

### Setup

We have java projects with maven as dependency manager. These java projects also have a `package.json` for yarn though to help with versioning and internal processes.  

The package.json looks something like this:
```json
{
  "name": "my-project",
  "version": "9.0.0",
  "description": "My project",
  "repository": "myrepo.com",
  "license": "UNLICENCED",
  "author": "Me",
  "scripts": {
    "bump-version": "cp-dev-utility bump-version",
    "update-dependencies": "cp-dev-utility update-dependencies"
  },
  "devDependencies": {
    "@mycompany/cp-dev-utility": "17.3.8"
  }
}
```

This "cp-dev-utility" tool is part of our private registry and should also be updated by renovate. Of course we provided authentication for that.
We also have a postUpgradeCommand that uses that tool, it looks something like that:
```js
    postUpgradeTasks: {
        "commands": ["yarn cp-dev-utility batch-bump-version changed"],
        "fileFilters": ["*", "**/helm-chart/**/*.yaml", "*/pom.xml", "*/src/local/resources/*"],
        "executionMode": "branch"
    },
```

### What works

If there is a new version of `cp-dev-utility`, renovate updates its version (and all maven dependencies) and executes the postUpgradeCommand as desired. All works fine!

### What does not work

If there is **no** new version of `cp-dev-utility,` then only the maven dependencies are updated but the postUpgradeCommand can not be executed because of this error:
![image](https://user-images.githubusercontent.com/84442685/161703936-0d80bf0c-fc21-49d8-a3ca-9b4a74540c34.png)

So therefore I tried to execute `yarn install` before `yarn cp-dev-utility batch-bump-version changed` and this results in the following error.
![yarn_install_error](https://user-images.githubusercontent.com/84442685/161704211-7842a099-bb4b-416c-87d1-f71882a627e3.jpg)

So is there any way to "force" the npm manager to login or to pass the token we defined in the `config.js` to the postUpgradeCommands somehow?

Help would be much appreciated! Thank you!

**Comment by rarkins** [comment, marked as answer]
Is the post upgrade command run after a maven update or a yarn/npm one?

**Reply by steknowis** [reply]
It should always run, also after a maven update.

**Reply by rarkins** [reply]
In that case setting `skipInstalls=false` won't help you, because there won't have been any `yarn` command run in the scenario of a maven update.

Another possibility could be utilizing `customEnvVariables`, i.e. you add your npm token there (note: via `secrets` so that it's sanitized). Then you could potentially `echo ..... >>` to a file or perhaps use env substitution in your yarn config file if yarn supports that

**Reply by steknowis** [reply]
Thank you very much for the quick and helpful response!

For those who might face a similar issue, we solved it like this:

```
module.exports = {

    secrets: {
        // NPM_TOKEN_READONLY is a GitLab variable
        NPM_TOKEN: process.env.NPM_TOKEN_READONLY,
    },

    customEnvVariables: {
        NPM_TOKEN: "{{ secrets.NPM_TOKEN }}",
    },

postUpgradeTasks: {
        "commands": [
            'echo "_auth=${NPM_TOKEN}" >> .npmrc',
            "yarn install --frozen-lockfile --non-interactive --production=false",
            "yarn cp-dev-utility batch-bump-version changed"
        ],
        "fileFilters": [
            // we must not commit .npmrc file as it contains a token - so allow all files that do not start with "."
            // and allow exceptions
            "*.*",
            ".gitlab-ci.yml",
            "*/pom.xml",
            "*/src/local/resources/*",
            "**/helm-chart/**/*.yaml"
        ],
        "executionMode": "branch"
    },
}
```
